controllers/periodic_healthcheck.go

```diff
@@ -0,0 +1,157 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"net/http"
+	"sigs.k8s.io/cluster-api/util/collections"
+	"strings"
+	"time"
+
+	etcdv1 "github.com/mrajashree/etcdadm-controller/api/v1alpha3"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/conditions"
+)
+
+const (
+	maxUnhealthyCount   = 5
+	healthCheckInterval = 30
+)
+
+type etcdHealthCheckConfig struct {
+	etcdHttpClient       *http.Client
+	etcdadmClusterMapper map[types.UID]etcdadmClusterMemberHealthConfig
+}
+
+type etcdadmClusterMemberHealthConfig struct {
+	unhealthyMembersFrequency map[string]int
+	unhealthyMembersToRemove  map[string]*clusterv1.Machine
+	endpointToMachineMapper   map[string]*clusterv1.Machine
+	cluster                   *clusterv1.Cluster
+	ownedMachines             collections.FilterableMachineCollection
+}
+
+func (r *EtcdadmClusterReconciler) startHealthCheckLoop(ctx context.Context, done <-chan struct{}) {
+	r.Log.Info("Starting periodic healthcheck loop")
+	ticker := time.NewTicker(healthCheckInterval * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-done:
+			return
+		case <-ticker.C:
+			etcdClusters := &etcdv1.EtcdadmClusterList{}
+			err := r.Client.List(ctx, etcdClusters)
+			if err != nil {
+				r.Log.Error(err, "Error listing etcdadm cluster objects")
+				continue
+			}
+			for _, ec := range etcdClusters.Items {
+				if !ec.Status.CreationComplete {
+					// etcdCluster not fully provisioned yet
+					continue
+				}
+				if conditions.IsFalse(&ec, etcdv1.EtcdMachinesSpecUpToDateCondition) {
+					// etcdCluster is undergoing upgrade, some machines might not be ready yet, skip periodic healthcheck
+					continue
+				}
+
+				var cluster *clusterv1.Cluster
+				if _, ok := r.etcdHealthCheckConfig.etcdadmClusterMapper[ec.UID]; !ok {
```

Comment:
does this map need to be in the etcdHealthCheckConfig object? since this runs in a subroutine you can run into race conditions with concurrent read/writes

Reply:
the only goroutine that reads/writes to this map is `startHealthCheckLoop` and it doesn't create any additional goroutines per etcdadmcluster, so I don't think this will lead to race conditions